File: Python/3-SlidingWindow/Problems/1-BestTime-toBuyandSellStock/BestTime-toBuyandSellStock.py
'''
Sliding Window: Best Time to Buy and Sell Stock

Problem Statement:
Given an array `prices` where `prices[i]` is the price of a given stock on the `i`th day,
return the maximum profit you can achieve from buying and selling one stock.

You may not engage in multiple transactions at the same time (i.e., you must sell before you buy again).
'''

# 1. Brute Force Approach (O(n^2))
def max_profit_brute_force(prices):
    """
    Calculates the maximum profit by trying every possible buy and sell day combination.
    It iterates through each day as a potential buying day and then iterates through all subsequent days
    as potential selling days to find the maximum profit.

    Args:
        prices (list): A list of stock prices for each day.

    Returns:
        int: The maximum profit that can be obtained.
    """
    max_profit = 0  # Initialize the maximum profit to 0.
    for i in range(len(prices)):  # Iterate through each day as a potential buying day.
        for j in range(i + 1, len(prices)):  # Iterate through each day after the buying day as a potential selling day.
            max_profit = max(max_profit, prices[j] - prices[i])  # Calculate the profit and update max_profit if it's larger.
    return max_profit  # Return the maximum profit found.

# 2. One Pass Sliding Window Approach (O(n))
def max_profit_sliding_window(prices):
    """
    Calculates the maximum profit using a sliding window approach with a single pass.  It keeps track of the minimum price
    seen so far and calculates the profit if sold on the current day.  The window implicitly expands as we go.

    Args:
        prices (list): A list of stock prices for each day.

    Returns:
        int: The maximum profit that can be obtained.
    """
    min_price = float('inf')  # Initialize the minimum price to infinity.
    max_profit = 0  # Initialize the maximum profit to 0.
    for price in prices:  # Iterate through the prices once.
        min_price = min(min_price, price)  # Update the minimum price if a lower price is found.
        max_profit = max(max_profit, price - min_price)  # Calculate the profit and update max_profit if it's larger.
    return max_profit  # Return the maximum profit.

# 3. Two Pointer Approach (O(n))
def max_profit_two_pointer(prices):
    """
    Calculates the maximum profit using the two-pointer approach.  One pointer represents the buying day,
    and the other represents the selling day.  It efficiently moves both pointers to find the optimal buy and sell days.

    Args:
        prices (list): A list of stock prices for each day.

    Returns:
        int: The maximum profit that can be obtained.
    """
    left, right = 0, 1  # Initialize the left pointer (buy) to the first day and the right pointer (sell) to the second day.
    max_profit = 0  # Initialize the maximum profit to 0.
    while right < len(prices):  # Iterate while the right pointer is within the price range.
        if prices[right] > prices[left]:  # If the selling price is greater than the buying price:
            max_profit = max(max_profit, prices[right] - prices[left])  # Calculate the profit and update max_profit.
        else:
            left = right  # If the selling price is not greater, move the buying day to the current selling day, as a lower price might be found later.
        right += 1  # Move the selling day to the next day.
    return max_profit  # Return the maximum profit.

# 4. Kadane’s Algorithm Adaptation (O(n))
def max_profit_kadane(prices):
    """
    Calculates the maximum profit by adapting Kadane's algorithm, which is typically used for finding the maximum subarray sum.
    Here, it's adapted to find the maximum difference between two prices in the array, representing the maximum profit.

    Args:
        prices (list): A list of stock prices for each day.

    Returns:
        int: The maximum profit that can be obtained.
    """
    max_profit, min_price = 0, prices[0]  # Initialize max_profit and min_price.  min_price starts at the first day's price.
    for price in prices[1:]:  # Iterate through prices starting from the *second* day.
        max_profit = max(max_profit, price - min_price)  # Calculate and update max_profit.
        min_price = min(min_price, price)  # Update min_price.
    return max_profit  # Return the maximum profit.

# 5. Monotonic Stack Approach (O(n))
def max_profit_monotonic_stack(prices):
    """
    Calculates the maximum profit using a monotonic stack.  The stack stores stock prices in increasing order.
    This ensures that for any price, the prices to its left in the stack are potential buying points with lower prices.

    Args:
        prices (list): A list of stock prices for each day.

    Returns:
        int: The maximum profit that can be obtained.
    """
    stack = []  # Initialize an empty stack.
    max_profit = 0  # Initialize max_profit.
    for price in prices:  # Iterate through the prices.
        # Maintain a monotonically increasing stack.  If the current price is lower than the top of the stack,
        # it means we've found a potentially better buying point, so pop the top of the stack.
        while stack and stack[-1] > price:
            stack.pop()
        # If the stack is not empty, calculate the profit between the current price and the lowest price so far (stack[0]).
        if stack:
            max_profit = max(max_profit, price - stack[0])
        stack.append(price)  # Push the current price onto the stack.
    return max_profit  # Return the maximum profit.

# Test Cases
prices = [7, 1, 5, 3, 6, 4]
print("Brute Force:", max_profit_brute_force(prices))
print("Sliding Window:", max_profit_sliding_window(prices))
print("Two Pointer:", max_profit_two_pointer(prices))
print("Kadane’s Algorithm:", max_profit_kadane(prices))
print("Monotonic Stack:", max_profit_monotonic_stack(prices))

File: C++/3-SlidingWindow/Problems/1-BestTime-toBuyandSellStock/BestTime-toBuyandSellStock.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits> // Required for numeric_limits

using namespace std;

// Function to find the maximum profit using the brute force approach.
// This approach has a time complexity of O(n^2) and a space complexity of O(1).
int maxProfitBruteForce(vector<int>& prices) {
    int maxProfit = 0;
    int n = prices.size();

    // Iterate through each possible buy day.
    for (int i = 0; i < n - 1; ++i) {
        // Iterate through each possible sell day after the buy day.
        for (int j = i + 1; j < n; ++j) {
            // Calculate the profit for the current buy and sell days.
            int profit = prices[j] - prices[i];
            // Update the maximum profit if the current profit is greater.
            maxProfit = max(maxProfit, profit);
        }
    }
    return maxProfit;
}

// Function to find the maximum profit using the optimized one-pass approach.
// This approach has a time complexity of O(n) and a space complexity of O(1).
int maxProfitOnePass(vector<int>& prices) {
    int minPrice = numeric_limits<int>::max(); // Initialize minPrice to the maximum possible integer value.
    int maxProfit = 0;
    int n = prices.size();

    // Iterate through the prices array only once.
    for (int i = 0; i < n; ++i) {
        // Update the minimum price if a lower price is found.
        minPrice = min(minPrice, prices[i]);
        // Calculate the profit if selling on the current day.
        int profit = prices[i] - minPrice;
        // Update the maximum profit if the current profit is greater.
        maxProfit = max(maxProfit, profit);
    }
    return maxProfit;
}

// Function to find the maximum profit using a modified one-pass approach
// This approach also has O(n) time complexity and O(1) space complexity.
int maxProfitOnePassOptimized(vector<int>& prices) {
    int maxProfit = 0;
    int buyPrice = prices[0]; // Assumes the first day is the initial buying day.
    int n = prices.size();

    for (int i = 1; i < n; ++i) {
        // If the current price is lower than the buying price, update the buying price.
        if (prices[i] < buyPrice) {
            buyPrice = prices[i];
        }
        // Otherwise, calculate the potential profit and update maxProfit.
        else {
            maxProfit = max(maxProfit, prices[i] - buyPrice);
        }
    }
    return maxProfit;
}

// Function to find the maximum profit using Kadane's Algorithm.
// Although Kadane's Algorithm is typically used for finding the maximum subarray sum,
// it can be adapted to solve this problem. This approach has a time complexity of O(n)
// and a space complexity of O(1).
int maxProfitKadane(vector<int>& prices) {
    int maxSoFar = 0;
    int currentMax = 0;
    int n = prices.size();

    for (int i = 1; i < n; ++i) {
        // Calculate the difference between consecutive days' prices.
        int diff = prices[i] - prices[i - 1];
        // Update currentMax.  If the difference is negative, currentMax will be 0.
        currentMax = max(0, currentMax + diff);
        // Update maxSoFar (the overall maximum profit).
        maxSoFar = max(maxSoFar, currentMax);
    }
    return maxSoFar;
}

// Function to find the maximum profit using dynamic programming.
// This approach has a time complexity of O(n) and a space complexity of O(n).  It's less
// space-efficient than the one-pass approaches, but illustrates dynamic programming.
int maxProfitDP(vector<int>& prices) {
    int n = prices.size();
    if (n == 0) return 0;

    // Create a DP table to store the maximum profit up to each day.
    vector<int> dp(n, 0);
    int minPrice = prices[0];

    // Iterate through the prices starting from the second day.
    for (int i = 1; i < n; ++i) {
        // Update the minimum price seen so far.
        minPrice = min(minPrice, prices[i]);
        // Calculate the maximum profit up to the current day.  It's either the
        // maximum profit up to the previous day, or the profit from buying at
        // the minimum price and selling today.
        dp[i] = max(dp[i - 1], prices[i] - minPrice);
    }
    // The last element of the DP table contains the overall maximum profit.
    return dp[n - 1];
}

int main() {
    // Example input: vector of stock prices.
    vector<int> prices = {7, 1, 5, 3, 6, 4};

    // Calculate and print the maximum profit using each approach.
    cout << "Maximum profit (Brute Force): " << maxProfitBruteForce(prices) << endl;
    cout << "Maximum profit (One Pass): " << maxProfitOnePass(prices) << endl;
    cout << "Maximum profit (One Pass Optimized): " << maxProfitOnePassOptimized(prices) << endl;
    cout << "Maximum profit (Kadane's Algorithm): " << maxProfitKadane(prices) << endl;
    cout << "Maximum profit (Dynamic Programming): " << maxProfitDP(prices) << endl;

    return 0;
}
